listeners/measured.go

```diff
@@ -3,38 +3,88 @@ package listeners
 import (
 	"net"
 	"net/http"
+	"sync"
 	"time"
 
 	"github.com/getlantern/measured"
 )
 
-// Wrapped stateAwareMeasuredListener that genrates the wrapped wrapMeasuredConn
+const (
+	rateInterval = 1 * time.Second
+)
+
+// MeasuredReportFN is a function that gets called to report stats from the
+// measured connection. deltaStats is like stats except that SentTotal and
+// RecvTotal are deltas relative to the prior reported stats. final indicates
+// whether this is the last call for a connection (i.e. connection has been
+// closed).
+type MeasuredReportFN func(ctx map[string]interface{}, stats *measured.Stats, deltaStats *measured.Stats,
+	final bool)
+
+// Wrapped stateAwareMeasuredListener that generates the wrapped wrapMeasuredConn
 type stateAwareMeasuredListener struct {
-	measured.MeasuredListener
+	net.Listener
+	reportInterval time.Duration
+	report         MeasuredReportFN
 }
 
-func NewMeasuredListener(l net.Listener, reportInterval time.Duration, m *measured.Measured) net.Listener {
-	return &stateAwareMeasuredListener{
-		MeasuredListener: *m.Listener(l, reportInterval),
-	}
+func NewMeasuredListener(l net.Listener, reportInterval time.Duration, report MeasuredReportFN) net.Listener {
+	return &stateAwareMeasuredListener{l, reportInterval, report}
 }
 
 func (l *stateAwareMeasuredListener) Accept() (c net.Conn, err error) {
-	c, err = l.MeasuredListener.Accept()
+	c, err = l.Listener.Accept()
 	if err != nil {
 		return nil, err
 	}
-	sac, _ := c.(*measured.Conn).Conn.(WrapConnEmbeddable)
-	return &wrapMeasuredConn{
-		WrapConnEmbeddable: sac,
-		Conn:               c.(*measured.Conn),
-	}, err
+	fs := make(chan *measured.Stats)
+	wc := &wrapMeasuredConn{
+		ctx:        make(map[string]interface{}),
+		finalStats: fs,
+		Conn: measured.Wrap(c, rateInterval, func(mc measured.Conn) {
+			fs <- mc.Stats()
+		}),
+		WrapConnEmbeddable: c.(WrapConnEmbeddable),
+	}
+	go wc.track(l.reportInterval, l.report)
+	return wc, nil
 }
 
 // Wrapped MeasuredConn that supports OnState
 type wrapMeasuredConn struct {
 	WrapConnEmbeddable
-	*measured.Conn
+	measured.Conn
+	ctx        map[string]interface{}
+	ctxMx      sync.RWMutex
+	finalStats chan *measured.Stats
+}
+
+func (c *wrapMeasuredConn) track(reportInterval time.Duration, report MeasuredReportFN) {
+	ticker := time.NewTicker(reportInterval)
+	var priorStats *measured.Stats
+	applyStats := func(final bool) {
+		stats := c.Conn.Stats()
+		deltaStats := stats
+		if priorStats != nil {
+			deltaStats.SentTotal -= priorStats.SentTotal
+			deltaStats.RecvTotal -= priorStats.RecvTotal
+		}
+		priorStats = stats
+		c.ctxMx.RLock()
+		ctx := c.ctx
+		c.ctxMx.RUnlock()
+		report(ctx, stats, deltaStats, final)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			applyStats(false)
+		case <-c.finalStats:
+			applyStats(true)
+			return
+		}
+	}
```

Comment:
With corresponding changes to `applyStats` of course, namely not calling `c.Conn.Stats` and accepting the stats as an argument.